PostSharp.Community.StructuralEquality.Weaver/EqualsInjection.cs

```diff
@@ -0,0 +1,408 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using PostSharp.Community.StructuralEquality.Weaver.Subroutines;
+using PostSharp.Sdk.CodeModel;
+using PostSharp.Sdk.CodeModel.Helpers;
+using PostSharp.Sdk.CodeModel.TypeSignatures;
+using PostSharp.Sdk.Extensibility;
+
+namespace PostSharp.Community.StructuralEquality.Weaver
+{
+    /// <summary>
+    /// Adds IEquatable&lt;T&gt; and implements equality methods.
+    /// </summary>
+    public class EqualsInjection
+    {
+        private readonly IGenericMethodDefinition referenceEqualsMethod;
+        private readonly IntrinsicTypeSignature objectType;
+        private readonly TypeDefDeclaration objectTypeDef;
+        private readonly ITypeSignature booleanType;
+        private readonly IGenericMethodDefinition getTypeMethod;
+        private readonly IGenericMethodDefinition instanceEqualsMethod;
+        private readonly IGenericMethodDefinition staticEqualsMethod;
+        private readonly IGenericMethodDefinition collectionEqualsMethod;
+        private readonly IGenericMethodDefinition getTypeFromHandleMethod;
+        private readonly ITypeSignature equatableInterface;
+
+        public EqualsInjection( Project project )
+        {
+            this.objectType = project.Module.Cache.GetIntrinsic( IntrinsicType.Object );
+            this.booleanType = project.Module.Cache.GetIntrinsic( IntrinsicType.Boolean );
+
+            this.objectTypeDef = this.objectType.GetTypeDefinition();
+
+            this.referenceEqualsMethod = project.Module.FindMethod( objectTypeDef, "ReferenceEquals" );
+            this.instanceEqualsMethod = project.Module.FindMethod( objectTypeDef, "Equals", 1 );
+            this.staticEqualsMethod = project.Module.FindMethod( objectTypeDef, "Equals", 2 );
+            this.getTypeMethod = project.Module.FindMethod( objectTypeDef, "GetType" );
+
+            var collectionHelperTypeDef = project.Module.Cache.GetType( typeof(CollectionHelper) ).GetTypeDefinition();
+            this.collectionEqualsMethod = project.Module.FindMethod( collectionHelperTypeDef, "Equals",
+                declaration => declaration.IsStatic );
+
+            var typeTypeDef = project.Module.Cache.GetType( typeof(Type) ).GetTypeDefinition();
+            this.getTypeFromHandleMethod = project.Module.FindMethod( typeTypeDef, "GetTypeFromHandle" );
+
+            this.equatableInterface = project.Module.Cache.GetType( typeof(IEquatable<>) );
+        }
+
+        public void AddEqualsTo( TypeDefDeclaration enhancedType, StructuralEqualityAttribute config,
+            ISet<FieldDefDeclaration> ignoredFields )
+        {
+            var typedEqualsMethod = this.InjectEqualsType( enhancedType, config, ignoredFields );

```

Comment:
In GetHashCode, I check if the GetHashCode method already exists. If it does, I don't overwrite it. We should probably decide on a common approach in this scenario.